File: backend/atlas/liveness/entropy.py
"""Entropy operators for liveness assessment (Math Spec v1.4 / GBSS).

The operators the spec names for scoring how *alive* a signal is: Shannon entropy,
Lempel-Ziv complexity, and spectral entropy — plus min-entropy (the conservative,
worst-case cousin we already use as a hard gate). Each takes raw sensor samples and
returns a scalar in a documented range.

LOAD-BEARING INVARIANT: these are MEASUREMENTS of liveness/freshness. Their outputs
only ever TIME and GATE the ratchet (feed the PoLE liveness gate); NONE is ever
folded into a key/value. The value stays clean QRNG. "Entropy times/refreshes; the
key is QRNG." A living body's entropy proves life — it is never the secret.

Pure Python (no numpy): the spectral operator uses a small direct DFT, fine for the
short per-window buffers used here.
"""

from __future__ import annotations

import math
from collections import Counter
from typing import Sequence


def shannon_entropy_bits(data: bytes) -> float:
    """Shannon entropy of the BYTE distribution, bits/byte (0..8). 0 for a constant;
    ~8 for uniform bytes. General diversity util."""
    return distribution_entropies(list(data))[0]


def distribution_entropies(symbols: Sequence) -> tuple[float, float]:
    """(Shannon, min-entropy) in bits over a sequence of hashable symbols. Shannon
    = average unpredictability (-Σ p·log2 p); min-entropy = worst-case
    (-log2 max p). Used across snapshots (each snapshot a symbol) so a bit-flipping
    replay loop still reads as few distinct symbols -> low entropy."""
    if not symbols:
        return 0.0, 0.0
    counts = Counter(symbols)
    n = len(symbols)
    shannon = 0.0
    p_max = 0.0
    for c in counts.values():
        p = c / n
        shannon -= p * math.log2(p)
        p_max = max(p_max, p)
    return shannon, -math.log2(p_max)


def lempel_ziv_complexity(data: bytes) -> float:
    """Normalized Lempel-Ziv (LZ76) complexity of the bit sequence, in ~[0,1].

    Counts the distinct patterns the sequence produces (Kaspar-Schuster), then
    normalizes by n/log2(n) — the asymptotic count for a random binary source. So
    genuine noise ~1.0 (incompressible), a repetitive / looped / constant stream ->
    low (compressible). Complements min-entropy for replay/loop detection: a loop
    that flips bits every tick still compresses well -> low LZ.
    """
    bits = "".join(f"{b:08b}" for b in data)
    n = len(bits)
    if n == 0:
        return 0.0
    i = 0
    c = 1           # number of distinct components
    ln = 1          # current prefix length being scanned
    k = 1           # length of the current match
    k_max = 1
    while ln + k <= n:
        if bits[i + k - 1] == bits[ln + k - 1]:
            k += 1
        else:
            k_max = max(k_max, k)
            i += 1
            if i == ln:                 # no match found starting anywhere in prefix
                c += 1
                ln += k_max
                i = 0
                k = 1
                k_max = 1
            else:
                k = 1
    if k != 1:                          # trailing partial component
        c += 1
    # normalize by the random-source asymptote b(n) = n/log2(n)
    norm = n / math.log2(n) if n > 1 else 1.0
    return c / norm


def spectral_entropy(waveform: Sequence[float]) -> float:
    """Normalized spectral entropy of a waveform, in [0,1]. Entropy of the power
    spectral density (DC removed) over its frequency bins, divided by log2(bins).

    Flat spectrum (white noise, a live broadband signal) -> ~1.0; a single dominant
    tone or a constant -> ~0. A live biological waveform (e.g. PPG) sits between —
    structured but not degenerate. Pure-Python direct DFT (O(n^2)); fine for short
    windows.
    """
    n = len(waveform)
    if n < 4:
        return 0.0
    mean = sum(waveform) / n
    x = [v - mean for v in waveform]            # remove DC so a constant -> 0
    half = n // 2
    psd = []
    for kf in range(1, half + 1):
        re = 0.0
        im = 0.0
        w = 2.0 * math.pi * kf / n
        for t in range(n):
            re += x[t] * math.cos(w * t)
            im -= x[t] * math.sin(w * t)
        psd.append(re * re + im * im)
    total = sum(psd)
    if total <= 0.0 or len(psd) < 2:
        return 0.0
    h = 0.0
    for power in psd:
        if power > 0:
            p = power / total
            h -= p * math.log2(p)
    return h / math.log2(len(psd))

File: ios/AtlasCore/Sources/AtlasCore/Liveness/Entropy.swift
import Foundation

/// Entropy operators for liveness assessment (Math Spec v1.4 / GBSS). Mirrors
/// `backend/atlas/liveness/entropy.py` byte-for-byte on the math: Shannon,
/// min-entropy, Lempel-Ziv complexity, spectral entropy.
///
/// LOAD-BEARING INVARIANT: these are MEASUREMENTS of liveness/freshness. Their
/// outputs only TIME and GATE the ratchet (feed the PoLE gate); NONE is ever folded
/// into a key/value. The value stays clean QRNG.
public enum Entropy {

    /// (Shannon, min-entropy) in bits over a sequence of symbols. Shannon = average
    /// unpredictability (-Σ p·log2 p); min-entropy = worst-case (-log2 max p).
    public static func distributionEntropies(_ symbols: [Data]) -> (shannon: Double, minEntropy: Double) {
        if symbols.isEmpty { return (0, 0) }
        var counts: [Data: Int] = [:]
        for s in symbols { counts[s, default: 0] += 1 }
        let n = Double(symbols.count)
        var shannon = 0.0, pMax = 0.0
        for c in counts.values {
            let p = Double(c) / n
            shannon -= p * log2(p)
            pMax = max(pMax, p)
        }
        return (shannon, -log2(pMax))
    }

    /// Shannon entropy of the byte distribution, bits/byte (0..8).
    public static func shannonBits(_ data: Data) -> Double {
        distributionEntropies(data.map { Data([$0]) }).shannon
    }

    /// Normalized Lempel-Ziv (LZ76) complexity of the bit sequence, ~[0,1]. Genuine
    /// noise ~1 (incompressible); a repetitive / looped / constant stream -> low.
    /// Complements min-entropy for replay/loop detection.
    public static func lempelZivComplexity(_ data: Data) -> Double {
        var chars: [Character] = []
        chars.reserveCapacity(data.count * 8)
        for b in data {
            let s = String(b, radix: 2)
            chars.append(contentsOf: String(repeating: "0", count: 8 - s.count) + s)
        }
        let n = chars.count
        if n == 0 { return 0.0 }
        var i = 0, c = 1, ln = 1, k = 1, kMax = 1
        while ln + k <= n {
            if chars[i + k - 1] == chars[ln + k - 1] {
                k += 1
            } else {
                kMax = max(kMax, k)
                i += 1
                if i == ln {
                    c += 1
                    ln += kMax
                    i = 0; k = 1; kMax = 1
                } else {
                    k = 1
                }
            }
        }
        if k != 1 { c += 1 }
        let norm = n > 1 ? Double(n) / log2(Double(n)) : 1.0
        return Double(c) / norm
    }

    /// Normalized spectral entropy of a waveform, [0,1]. Entropy of the power
    /// spectral density (DC removed) over its bins, /log2(bins). Flat spectrum
    /// (broadband/live) -> ~1; a single tone or constant -> ~0. Pure direct DFT.
    public static func spectralEntropy(_ waveform: [Double]) -> Double {
        let n = waveform.count
        if n < 4 { return 0.0 }
        let mean = waveform.reduce(0, +) / Double(n)
        let x = waveform.map { $0 - mean }
        let half = n / 2
        var psd: [Double] = []
        for kf in 1...half {
            var re = 0.0, im = 0.0
            let w = 2.0 * Double.pi * Double(kf) / Double(n)
            for t in 0..<n {
                re += x[t] * cos(w * Double(t))
                im -= x[t] * sin(w * Double(t))
            }
            psd.append(re * re + im * im)
        }
        let total = psd.reduce(0, +)
        if total <= 0 || psd.count < 2 { return 0.0 }
        var h = 0.0
        for power in psd where power > 0 {
            let p = power / total
            h -= p * log2(p)
        }
        return h / log2(Double(psd.count))
    }
}
